### PR TITLE
Upgrade websocket failure add extra error info

### PR DESCRIPTION
**Pull request by seans3**
* For WebSocket upgrade failure add extra error information from the response body (if it exists).
  * A similar dynamic already happens for the SPDY upgrade failure case.

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/transport/websocket/roundtripper.go [hunk 1]
```diff
@@ -111,9 +113,14 @@ func (rt *RoundTripper) RoundTrip(request *http.Request) (retResp *http.Response
 	wsConn, resp, err := dialer.DialContext(request.Context(), request.URL.String(), request.Header)
 	if err != nil {
 		if errors.Is(err, gwebsocket.ErrBadHandshake) {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/transport/websocket/roundtripper.go:115]: Mimic what SpdyRoundTripper#NewConnection does more closely, trying to read the response into a status error, and not dumping it directly into the output if it is a status error. Something like:

```go
if resp != nil && len(resp.Status) > 0 {
		defer resp.Body.Close()
    var cause error
		responseError := ""
		responseErrorBytes, readErr := io.ReadAll(io.LimitReader(resp.Body, 64*1024))
		if readErr != nil {
			cause = fmt.Errorf("%w (unable to read error from server response)", err)
		} else {
			responseError = strings.TrimSpace(string(responseErrorBytes))
			if obj, _, err := statusCodecs.UniversalDecoder().Decode(responseErrorBytes, nil, &metav1.Status{}); err == nil {
				if status, ok := obj.(*metav1.Status); ok {
					cause = &apierrors.StatusError{ErrStatus: *status}
				}
			}
			if cause == nil {
				cause = fmt.Errorf("%w: %s", err, responseError)
			}
		}
		return nil, &httpstream.UpgradeFailureError{Cause: cause}
```

**Comment by aojea** [comment]
> pull-kubernetes-integration — Job failed.                     BaseSHA:6df3b28f22c3984fa5a35267dd1bf47129ef5b93

https://github.com/kubernetes/kubernetes/issues/127830

/lgtm

**Review by liggitt — COMMENTED** [review]
one nit, then lgtm

File: staging/src/k8s.io/client-go/transport/websocket/roundtripper.go [hunk 2]
```diff
@@ -110,12 +127,33 @@ func (rt *RoundTripper) RoundTrip(request *http.Request) (retResp *http.Response
...
+			cause := err
+			// Enhance the error message with the error response if possible.
 			if resp != nil && len(resp.Status) > 0 {
-				err = fmt.Errorf("%w (%s)", err, resp.Status)
+				defer resp.Body.Close()                         //nolint:errcheck
+				cause = fmt.Errorf("%w (%s)", err, resp.Status) // Always add the response status
+				responseError := ""
+				responseErrorBytes, readErr := io.ReadAll(io.LimitReader(resp.Body, 64*1024))
+				if readErr != nil {
+					cause = fmt.Errorf("%w (unable to read error from server response)", cause)
```
[the remark this answers is not in the document]
Comment (seans3) [reply staging/src/k8s.io/client-go/transport/websocket/roundtripper.go]: Fixed.

**Pull request merged**